### PR TITLE
OTLP receiver: Generate `target_info` samples between the earliest and latest samples per resource

### PR DESCRIPTION
**Pull request by aknuds1**
Modify the OTLP receiver to generate `target_info` samples starting at the earliest and ending at the latest samples per resource, instead of only one for the latest timestamp. Between those samples, one sample is generated every lookback delta/2 timestamps. The idea behind generating `target_info` samples spaced lookback delta/2 apart is to ensure that `target_info` is found for every connected time series sample generated from the same request.

The reason for adding a sample corresponding to the latest timestamp is that otherwise you may miss the latest `target_info sample` when doing a join or `info` function query, if you have to look back more than lookback delta/2 for the latest corresponding non-info sample.

The motivation behind this fix is that I realized when debugging strange `info` PromQL function behaviour that the reason was that `target_info` started one minute later than the time series it was being joined with (by `info`). See [Grafana Mimir issue](https://github.com/grafana/mimir/issues/11756) for reference.

### CODE REVIEW

**Review by beorn7 — COMMENTED** [review]
Why do we need a sample coinciding with the latest timestamp?

Having one at the earliest timestamp and then one at half the lookback delta until we reach or exceed the latest timestamp should be enough, shouldn't it?

**Comment by aknuds1** [comment]
> Why do we need a sample coinciding with the latest timestamp?
>
> Having one at the earliest timestamp and then one at half the lookback delta until we reach or exceed the latest timestamp should be enough, shouldn't it?

@beorn7 My algorithm doesn't exceed the latest timestamp, it ends exactly at the latest timestamp. Is there any benefit to modify it, so it instead potentially exceeds the latest timestamp? I don't see it at least.

**Comment by beorn7** [comment]
What I mean is to _not_ add a sample that exceed the latest timestamp, but also don't add any sample that doesn't match the lookbackdelta/2 repeating pattern.

Just add one at the beginning of the range, and then every lookbackdelta/2 duration, until you have added the last sample _before_ the end of the range.

**Comment by aknuds1** [comment]
> What I mean is to not add a sample that exceed the latest timestamp, but also don't add any sample that doesn't match the lookbackdelta/2 repeating pattern.

@beorn7 Isn't the problem with this approach that you can miss the latest `target_info` sample when doing a join or `info` function query, if you have to look back more than lookback delta/2 for the latest corresponding non-info sample?

**Comment by beorn7** [comment]
Oh yes, you are right. Now I got it.

(Maybe it should be explained in a way so that even people like me will understand it. :)

So this looks plausible to me, then. I'll leave the codelevel review to the OTLP receiver experts.

**Comment by aknuds1** [comment]
> (Maybe it should be explained in a way so that even people like me will understand it. :)

@beorn7 Hahahaha! :) I can update the PR description to be explicit about this, thanks for the feedback!

File: storage/remote/otlptranslator/prometheusremotewrite/metrics_to_prw.go [hunk 1]
```diff
@@ -264,7 +269,9 @@ func (c *PrometheusConverter) FromMetrics(ctx context.Context, md pmetric.Metric
 				}
 			}
 		}
-		addResourceTargetInfo(resource, settings, mostRecentTimestamp, c)
+		if earliestTimestamp < pcommon.Timestamp(math.MaxUint64) && latestTimestamp > 0 {
```
Comment (fionaliao) [inline storage/remote/otlptranslator/prometheusremotewrite/metrics_to_prw.go]: Nit: Is it valid to have samples with ts = 0? If so, we could get in a case where a valid sample with ts 0 is written but doesn't have a target_info. In real scenarios, having ts = 0 is probably unlikely though

**Review by fionaliao — APPROVED** [review]
Logic makes sense to me 👍

File: storage/remote/otlptranslator/prometheusremotewrite/metrics_to_prw.go (hunk 1 above)
Comment (aknuds1) [reply storage/remote/otlptranslator/prometheusremotewrite/metrics_to_prw.go]: Taking a look, thx.

**Review by ArthurSens — APPROVED** [review]
LGTM!

File: storage/remote/otlptranslator/prometheusremotewrite/metrics_to_prw.go (hunk 1 above)
Comment (aknuds1) [reply storage/remote/otlptranslator/prometheusremotewrite/metrics_to_prw.go]: It's a good point you're raising @fionaliao. I thought it wouldn't make sense to have samples starting at Unix epoch (0), but maybe it's better to not make that assumption? In retrospect, I think it's enough to just check that `earliestTimestamp < pcommon.Timestamp(math.MaxUint64)`. Checking also `latestTimestamp` is overkill.

Comment (aknuds1) [reply storage/remote/otlptranslator/prometheusremotewrite/metrics_to_prw.go]: I dropped the check for `latestTimestamp > 0` @fionaliao, since I think it's overkill as explained above. Thanks!

Comment (fionaliao) [reply storage/remote/otlptranslator/prometheusremotewrite/metrics_to_prw.go]: Thanks for updating that. Though I agree that there aren't many cases where having samples at unix epoch makes a lot of sense. 

It could possibly could be a minor issue in testing - there are a few places in the promql tests where we load samples at ts 0 for example. So there could be a case in the future where a test is added for the otlp endpoint and tries to use zero as a legitimate timestamp and then we hit a weird edge case, but that's really reaching.
